### PR TITLE
Adding template & var file names to bake keys.

### PR DESCRIPTION
**Pull request by varikin**
Helps prevent bakes with different files stomp on each other.

### CODE REVIEW

**Comment by varikin** [comment]
@spinnaker/reviewers PTAL

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy [hunk 1]
```diff
@@ -71,30 +71,39 @@ abstract class CloudProviderBakeHandler {
    * bake.
    */
   String produceBakeKey(String region, BakeRequest bakeRequest) {
-    // TODO(duftler): Consider whether template_file_name, extended_attributes and var_file_name should be
-    // taken into account when composing the bake key.
     bakeRequest.with {
-      def bakeKey = "bake:$cloud_provider_type:$base_os"
+      def keys = [
+              'bake',
```
Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy]: Indentation is off here.

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy [hunk 2]
```diff
@@ -71,30 +71,39 @@ abstract class CloudProviderBakeHandler {
...
+              cloud_provider_type,
+              base_os
+      ]
 
       if (base_ami) {
-        bakeKey += ":$base_ami"
+        keys << base_ami
```
Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy]: Now that there are so many optional key segments, maybe it makes sense to extract an `addIfPresent()` utility method?

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy [hunk 3]
```diff
@@ -749,4 +749,23 @@ class OpenstackBakeHandlerSpec extends Specification implements TestDefaults {
       bakeKey1 == "bake:openstack:centos:kato|nflx-djangobase-enhanced_0.1-h12.170cdbd_all|mongodb:RegionOne"
       bakeKey2 == bakeKey1
   }
+
+  def 'produce an openstack bakeKey with a template and var file names'() {
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: The indentation in this test does not match its siblings.

Comment (varikin) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: Did I get all the indentation problems?

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy [hunk 4]
```diff
@@ -71,30 +71,26 @@ abstract class CloudProviderBakeHandler {
...
-
+      def keys = [
+        'bake',
+        cloud_provider_type,
+        base_os,
+        base_ami,
+        ami_name,
+        template_file_name,
+        var_file_name,
+      ].findAll { it }
```
Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy]: Ah, even nicer.

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy [hunk 5]
```diff
@@ -71,30 +71,26 @@ abstract class CloudProviderBakeHandler {
...
+
+      // Package name is always part of key, even if it is an empty string
+      keys << (package_name ? package_name.tokenize().join('|') : "")
```
Comment (albertofem) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy]: Shouldn't this be access from `bakeRequest`. If not, this variable is empty and the package will be always `""`, right?

**Pull request merged**